### PR TITLE
test: Add addrman peers.dat test

### PR DESCRIPTION
**Pull request by maflcko**

### CODE REVIEW

**Comment by MarcoFalke** [comment]
@jonatack https://github.com/bitcoin/bitcoin/pull/22762#issuecomment-912441771

File: test/functional/feature_addrman.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        self.log.info("Check that addrman from future can not be read")
+        self.stop_node(0)
+        write_addrman(peers_dat, lowest_compatible=111)
+        with self.nodes[0].assert_debug_log([
+                f'ERROR: DeserializeDB: Deserialize or I/O error - Unsupported format of addrman database: \\x01. It is compatible with formats >=111, but the maximum supported by this version of {self.config["environment"]["PACKAGE_NAME"]} is 3.: iostream error',
+                'Recreating peers.dat',
+        ]):
+            self.start_node(0)
+
+        self.log.info("Check that corrupt addrman can not be read")
```
Comment (jonatack) [inline test/functional/feature_addrman.py:61]: nit, s/can not/cannot/ here and line 52

File: test/functional/feature_addrman.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        self.num_nodes = 1
+
+    def run_test(self):
+        peers_dat = os.path.join(self.nodes[0].datadir, self.chain, "peers.dat")
+
+        self.log.info("Check that mocked addrman is valid")
+        self.stop_node(0)
+        write_addrman(peers_dat)
+        with self.nodes[0].assert_debug_log(['Loaded 0 addresses from peers.dat']):
+            self.start_node(0)
```
Comment (jonatack) [inline test/functional/feature_addrman.py:50]:
```suggestion
            self.start_node(0, ["-checkaddrman=1"])
```
can also check that `self.nodes[0].getnodeaddresses()` is an empty array here and in the last test, to not rely only on the log

**Review by jonatack — COMMENTED** [review]
Nice, concept ACK. Some initial ideas below.

File: test/functional/feature_addrman.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        self.stop_node(0)
+        with open(peers_dat, 'wb') as f:
+            f.write(serialize_addrman()[:-1])
+        with self.nodes[0].assert_debug_log([
+                'ERROR: DeserializeDB: Deserialize or I/O error - CAutoFile::read: end of file: iostream error',
+                'Recreating peers.dat',
+        ]):
+            self.start_node(0)
+
+        self.log.info("Check that missing addrman is recreated")
```
Comment (jonatack) [inline test/functional/feature_addrman.py:71]: This is the same as [`d507068` (#22831)](https://github.com/bitcoin/bitcoin/pull/22831/commits/d5070688feb68fac59b076411f1d84131726098e)

File: test/functional/feature_addrman.py (hunk 1 above)
Comment (MarcoFalke) [reply test/functional/feature_addrman.py]: Thanks, done

File: test/functional/feature_addrman.py (hunk 2 above)
Comment (MarcoFalke) [reply test/functional/feature_addrman.py]: Thanks, done

File: test/functional/feature_addrman.py (hunk 3 above)
Comment (MarcoFalke) [reply test/functional/feature_addrman.py:75]: Thanks for the pointer

**Comment by MarcoFalke** [comment]
Force pushed to address nits

**Comment by jonatack** [comment]
I wrote some follow-ups last weekend to #22831 to add more addrman tests and rename `feature_asmap.py` to `feature_addrman_asmap.py` as they are related features, but didn't want to add them to that pull to keep it focused on just fixing the bug with regression tests.

**Comment by MarcoFalke** [comment]
Looks like there is some kind of compiler bug or compiler behaviour change

**Comment by MarcoFalke** [comment]
Reported issue in https://github.com/c42f/tinyformat/issues/82, worked around issue in #22879

**Pull request closed without merging**